src/commands/serve.rs

```diff
@@ -0,0 +1,201 @@
+use anyhow::Result;
+use clap::Parser;
+use std::{path::PathBuf, pin::Pin, sync::Arc};
+use wasmtime::component::{Component, InstancePre, Linker};
+use wasmtime::{Engine, Store};
+use wasmtime_cli_flags::CommonOptions;
+use wasmtime_wasi::preview2::{Table, WasiCtx, WasiCtxBuilder, WasiView};
+use wasmtime_wasi_http::{body::HyperOutgoingBody, WasiHttpCtx, WasiHttpView};
+
+struct Host {
+    table: Table,
+    ctx: WasiCtx,
+    http: WasiHttpCtx,
+}
+
+impl Host {
+    fn new() -> Result<Self> {
+        let mut table = Table::new();
+        let ctx = WasiCtxBuilder::new().build(&mut table)?;
+        Ok(Host {
+            table,
+            ctx,
+            http: WasiHttpCtx,
+        })
+    }
+}
+
+impl WasiView for Host {
+    fn table(&self) -> &Table {
+        &self.table
+    }
+
+    fn table_mut(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&self) -> &WasiCtx {
+        &self.ctx
+    }
+
+    fn ctx_mut(&mut self) -> &mut WasiCtx {
+        &mut self.ctx
+    }
+}
+
+impl WasiHttpView for Host {
+    fn table(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&mut self) -> &mut WasiHttpCtx {
+        &mut self.http
+    }
+}
+
+/// Runs a WebAssembly module
+#[derive(Parser)]
+#[structopt(name = "run")]
+pub struct ServeCommand {
+    #[clap(flatten)]
+    common: CommonOptions,
+
+    /// Socket address for the web server to bind to. Defaults to 0.0.0.0:8080.
+    #[clap(long = "addr", value_name = "SOCKADDR")]
+    addr: Option<std::net::SocketAddr>,
+
+    /// The WebAssembly component to run.
+    #[clap(value_name = "WASM", required = true)]
+    component: PathBuf,
+}
+
+impl ServeCommand {
+    fn addr(&self) -> std::net::SocketAddr {
+        self.addr.unwrap_or("0.0.0.0:8080".parse().unwrap())
+    }
+
+    /// Start a server to run the given wasi-http proxy component
+    pub fn execute(mut self) -> Result<()> {
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_time()
+            .enable_io()
+            .build()?;
+
+        runtime.block_on(async move {
+            tokio::select! {
+                _ = tokio::signal::ctrl_c() => {
+                    Ok::<_, anyhow::Error>(())
+                }
```

Comment:
Out of curiosity is it easy enough to have a graceful shutdown here where ctrl-c sets a flag "please exit ASAP"? Otherwise this looks like it cancel all pending work if it's happening during ctrl-c (in which case this probably isn't much better than avoiding installing a ctrl-c in the first place)

Reply:
Yeah, I think that should be possible: we could share some state with the running thread that we check for each iteration of the accept loop.

Reply:
I'm going to punt this to a follow-up PR, in the interest of landing this so that the resourcification work can proceed.

Reply:
https://github.com/bytecodealliance/wasmtime/issues/7107